src/app/warehouse/warehouse.hpp

```diff
@@ -0,0 +1,26 @@
+#pragma once
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <vector>
+
+class Product;
+
+class Warehouse
+{
+  public:
+    void increaseAmount(const std::string &name, const uint amount);
+    void decreaseAmount(const std::string &name, const uint amount);
+    void addProducts(const std::vector<std::shared_ptr<Product>> &products);
+    bool validateAdd(const std::string &name) const;
+    void removeProducts(const std::vector<std::string> &products);
+    bool validateRemoval(const std::string &name) const;
+    std::shared_ptr<Product> getPtrToProduct(const std::string &name);
+    std::vector<std::shared_ptr<Product>> &getExpiredProducts(); // TODO: write this function when type Date will be
+                                                                 // created
+    std::vector<std::shared_ptr<Product>> &getProducts();
+
+  private:
+    bool is_valid_state;
+    std::vector<std::shared_ptr<Product>> products_;
```

Comment:
sed::set is better suited for our needs to store associations.

Reply:
std::set seems ritght at first glance, but might lead to storing more than one Product of the same type in the container (which might be a good idea itself- please read further remarks). std::set uses operator < for comparison but on addresses.

This leeds me to wonder if we even want to store shared_ptrs or objects.
Maybe storing Product object will be enough?
It depends mainly on the question if we assume existance of more than one Product object in container.
1. we can add Product by copying it or moving it our container.
2. we can return a raw pointer to our Product object stored in the vector or vector<*> in getProducts() case
3. we return nullptr from getProduct() if we don't find it anyway.
After all it seems, that a Warehouse will be the sole owner of The product.

In such case we could consider std::set again but it is not obvious for me if it is a good idea:
1. we would have to add operator< for Product (not a great problem after all)
2. we would have to check if product is already stored any way in order not to override an existing one (thus removing all data)

**Most important:** we should decide if we want to store **more objects of the same Product name**?
At first glance the answer is "no". But if we have for example same Medicine but with different expiration dates (one already stored may have shorter one than one added to the Warehouse). In such case we can not simply add them in one object.
I think that the answer to this question is **crucial** as it will impact implementation of many Warehouse member functions (if I analysed it right - currently we predict only one Product object stored).

Reply:
Ok, I'm doing it with std::set

Reply:
Because we want store more objects of the same Product name - I choose to use std::vector instead of std::set.

Reply:
Done